diff --git a/lib/rspec/core/metadata_filter.rb b/lib/rspec/core/metadata_filter.rb
index 2e63baf9..b7ab008c 100644
--- a/lib/rspec/core/metadata_filter.rb
+++ b/lib/rspec/core/metadata_filter.rb
@@ -234,7 +234,17 @@ module RSpec
 
         def proc_keys_from(metadata)
           metadata.each_with_object([]) do |(key, value), to_return|
-            to_return << key if Proc === value
+            to_return << key if has_proc?(value)
+          end
+        end
+
+        def has_proc?(value)
+          if Proc === value
+            true
+          elsif Hash === value
+            value.values.any? { |v| has_proc?(v) }
+          else
+            false
           end
         end
 
@@ -243,7 +253,7 @@ module RSpec
           undef proc_keys_from
           def proc_keys_from(metadata)
             metadata.inject([]) do |to_return, (key, value)|
-              to_return << key if Proc === value
+              to_return << key if has_proc?(value)
               to_return
             end
           end
diff --git a/spec/rspec/core/filterable_item_repository_spec.rb b/spec/rspec/core/filterable_item_repository_spec.rb
index 99a526f9..d011c2a5 100644
--- a/spec/rspec/core/filterable_item_repository_spec.rb
+++ b/spec/rspec/core/filterable_item_repository_spec.rb
@@ -98,6 +98,19 @@ module RSpec
                 end
               end
 
+              context "with proc values in nested hash" do
+                before do
+                  add_item item_4, { :nested => { :include_it => flip_proc } }
+                end
+
+                it 'evaluates the proc each time since the logic can return a different value each time' do
+                  expect(repo.items_for(:nested => { :include_it => nil })).to contain_exactly(item_4)
+                  expect(repo.items_for(:nested => { :include_it => nil })).to eq([])
+                  expect(repo.items_for(:nested => { :include_it => nil })).to contain_exactly(item_4)
+                  expect(repo.items_for(:nested => { :include_it => nil })).to eq([])
+                end
+              end
+
               context "when initialized with the `:any?` predicate" do
                 let(:repo) { FilterableItemRepository::QueryOptimized.new(:any?) }
 
@@ -222,6 +235,53 @@ module RSpec
               end
             end
 
+            context "when there are some nested proc keys" do
+              before do
+                add_item item_4, { :nested => { :include_it => flip_proc } }
+              end
+
+              it 'still performs memoization for metadata hashes that lack those keys' do
+                call_counts = track_metadata_filter_apply_calls
+
+                expect(repo.items_for(:slow => true, :other => "foo")).to contain_exactly(item_2)
+                expect(repo.items_for(:slow => true, :other => "foo")).to contain_exactly(item_2)
+
+                expect(call_counts[:slow => true]).to eq(1)
+              end
+            end
+
+            context "when there are nested hash keys without procs" do
+              before do
+                add_item item_4, { :nested => { :include_it => true } }
+              end
+
+              it 'performs memoization for metadata hashes' do
+                call_counts = track_metadata_filter_apply_calls
+
+                expect(repo.items_for(:nested => { :include_it => true })).to contain_exactly(item_4)
+                expect(repo.items_for(:nested => { :include_it => true })).to contain_exactly(item_4)
+
+                expect(call_counts[:nested => { :include_it => true }]).to eq(1)
+              end
+            end
+
+            describe "reconstructing proc keys" do
+              it "memoizes properly after deleting a nested proc key" do
+                add_item item_4, { :nested => { :include_it => flip_proc } }
+                expect(repo.items_for(:nested => { :include_it => nil })).to contain_exactly(item_4)
+                
+                repo.delete(item_4, { :nested => { :include_it => flip_proc } })
+                
+                add_item item_4, { :nested => { :include_it => true } }
+                
+                call_counts = track_metadata_filter_apply_calls
+                expect(repo.items_for(:nested => { :include_it => true })).to contain_exactly(item_4)
+                expect(repo.items_for(:nested => { :include_it => true })).to contain_exactly(item_4)
+                
+                expect(call_counts[:nested => { :include_it => true }]).to eq(1)
+              end
+            end
+
             def track_metadata_filter_apply_calls
               Hash.new(0).tap do |call_counts|
                 allow(MetadataFilter).to receive(:apply?).and_wrap_original do |original, predicate, item_meta, request_meta|
