diff --git a/lib/rspec/core/option_parser.rb b/lib/rspec/core/option_parser.rb
index e239e87a..b8fb734c 100644
--- a/lib/rspec/core/option_parser.rb
+++ b/lib/rspec/core/option_parser.rb
@@ -229,7 +229,11 @@ FILTERING
 
         parser.on('--exclude-pattern PATTERN',
                   'Load files except those matching pattern. Opposite effect of --pattern.') do |o|
-          options[:exclude_pattern] = o
+          if options[:exclude_pattern]
+            options[:exclude_pattern] += ',' + o
+          else
+            options[:exclude_pattern] = o
+          end
         end
 
         parser.on('-e', '--example STRING', "Run examples whose full nested names include STRING (may be",
diff --git a/spec/rspec/core/configuration_options_spec.rb b/spec/rspec/core/configuration_options_spec.rb
index 4bf93974..f94d0243 100644
--- a/spec/rspec/core/configuration_options_spec.rb
+++ b/spec/rspec/core/configuration_options_spec.rb
@@ -199,6 +199,22 @@ RSpec.describe RSpec::Core::ConfigurationOptions, :isolated_directory => true, :
       end
     end
 
+    it "overrides --pattern when specified by multiple configuration sources" do
+      with_env_vars 'SPEC_OPTS' => "--pattern env_pattern" do
+        opts = config_options_object(*%w[--pattern opts_pattern])
+        expect(config).to receive(:force).with({:pattern => "env_pattern"}).ordered
+        opts.configure(config)
+      end
+    end
+
+    it "overrides --exclude-pattern when specified by multiple configuration sources" do
+      with_env_vars 'SPEC_OPTS' => "--exclude-pattern env_pattern" do
+        opts = config_options_object(*%w[--exclude-pattern opts_pattern])
+        expect(config).to receive(:force).with({:exclude_pattern => "env_pattern"}).ordered
+        opts.configure(config)
+      end
+    end
+
     %w[ --only-failures --next-failure -n].each do |option|
       describe option do
         it "changes `config.only_failures?` to true" do
diff --git a/spec/rspec/core/option_parser_spec.rb b/spec/rspec/core/option_parser_spec.rb
index af3edf4e..568f6e44 100644
--- a/spec/rspec/core/option_parser_spec.rb
+++ b/spec/rspec/core/option_parser_spec.rb
@@ -240,6 +240,24 @@ module RSpec::Core
       end
     end
 
+    describe '--exclude-pattern' do
+      it "sets the filename pattern" do
+        options = Parser.parse(['--exclude-pattern', 'spec/**/*.spec'])
+        expect(options[:exclude_pattern]).to eq('spec/**/*.spec')
+      end
+
+      it 'combines multiple patterns' do
+        options = Parser.parse(['--exclude-pattern', 'spec/**/*.spec', '--exclude-pattern', 'tests/**/*.spec'])
+        expect(options[:exclude_pattern]).to eq('spec/**/*.spec,tests/**/*.spec')
+      end
+
+      it 'maintains independent values from --pattern' do
+        options = Parser.parse(['--pattern', 'a/**/*.spec', '--exclude-pattern', 'b/**/*.spec', '--pattern', 'c/**/*.spec', '--exclude-pattern', 'd/**/*.spec'])
+        expect(options[:pattern]).to eq('a/**/*.spec,c/**/*.spec')
+        expect(options[:exclude_pattern]).to eq('b/**/*.spec,d/**/*.spec')
+      end
+    end
+
     %w[--tag -t].each do |option|
       describe option do
         context "without ~" do
